clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/controllers/GoogleNamedImageLookupController.java

```diff
@@ -0,0 +1,207 @@
+
+    // Set of accounts for which we should return images: either the supplied account, or default
+    // to all accounts
+    Set<String> accounts;
+    if (account != null && !account.equals("")) {
+      accounts = new HashSet<>();
+      if (imageMap.containsKey(account)) {
+        accounts.add(account);
+      }
+    } else {
+      accounts = imageMap.keySet();
+    }
+
+    List<NamedImage> results = new ArrayList<>();
+    for (String imageAccount : accounts) {
+      for (Image i : imageMap.get(imageAccount)) {
+        Map<String, Object> attributes = new HashMap<>();
+        attributes.put("creationDate", i.get("creationTimestamp"));
+        NamedImage newImage = new NamedImage(imageAccount, i.getName(), attributes, buildTagsMap(i));
+        results.add(newImage);
+      }
+    }
+
+    Predicate<NamedImage> queryFilter = i -> true;
+    if (q != null && q.trim().length() > 0) {
+      String glob = q.trim();
+      // Wrap in '*' if there are no glob-style characters in the query string.
+      if (!glob.contains("*") && !glob.contains("?") && !glob.contains("[") && !glob.contains("\\")) {
+        glob = "*" + glob + "*";
+      }
+      Pattern pattern = new InMemoryCache.Glob(glob).toPattern();
+      queryFilter = i-> pattern.matcher(i.imageName).matches();
+    }
+
+    return results.stream()
+      .filter(queryFilter)
+      .filter(namedImage -> matchesTagFilters(namedImage, extractTagFilters(request)))
+      .sorted(Comparator.comparing(image -> image.imageName))
+      .collect(Collectors.toList());
+  }
+
+  private Map<String, List<Image>> listImagesByAccount() {
+    Collection<String> identifiers = cacheView.filterIdentifiers(IMAGES.getNs(), GoogleCloudProvider.getID() + ":*");
+    Map<String, List<Image>> result = new HashMap<>();
+
+    Collection<CacheData> allCacheData = cacheView.getAll(IMAGES.getNs(), identifiers, RelationshipCacheFilter.none());
+    allCacheData.forEach(
+      cacheData -> {
+        String account = Keys.parse(cacheData.getId()).get("account");
+        if (!result.containsKey(account)) {
+          result.put(account, new ArrayList<>());
+        }
+        Object hashImage = cacheData.getAttributes().get("image");
+        try {
+          Image myImage = jsonMapper.fromString(objectMapper.writeValueAsString(hashImage), Image.class);
+          result.get(account).add(myImage);
+        } catch (IOException e) {
+          throw new RuntimeException("Image deserialization failed");
+        }
+      }
+    );
+
+    return result;
+  }
+
+  @VisibleForTesting
+  public static Map<String, String> buildTagsMap(Image image) {
+    Map<String, String> tags = new HashMap<>();
+
+    String description = image.getDescription();
+    if (description != null) {
+      tags = Arrays.stream(description.split(","))
+        .filter(token -> token.contains(": "))
+        .map(token -> token.split(": ", 2))
+        .collect(Collectors.toMap(
+          token -> token[0].trim(),
```

Comment:
This can throw if `token[0].trim()` ever produces a duplicate key. Are you sure that can't happen? If not, supply some sort of merge function, as simple as `(a, b) -> a`